onnxruntime/core/graph/graph.cc

```diff
@@ -2917,10 +2920,28 @@ Status Graph::ReplaceInitializedTensor(const ONNX_NAMESPACE::TensorProto& new_in
   ORT_ENFORCE(existing_entry != mutable_initializers.pointer_end(),
               "graph_proto_ is not in sync with name_to_initial_tensor_");
 
-  **existing_entry = new_initializer;
+  **existing_entry = std::move(new_initializer);
+
+  return Status::OK();
+}
+
+Status Graph::ReplaceInitializedTensor(ONNX_NAMESPACE::TensorProto new_initializer) {
+  return ReplaceInitializedTensorImpl(std::move(new_initializer), false);
+}
 
+#if !defined(DISABLE_EXTERNAL_INITIALIZERS)
+Status Graph::InjectExternalInitializedTensors(const InlinedHashMap<std::string, OrtValue>& external_initializers) {
+  for (const auto& e : external_initializers) {
+    const auto& name = e.first;
+    const OrtValue& ort_value = e.second;
+    auto tensor_proto = utils::TensorToTensorProto(ort_value.Get<Tensor>(), name);
```

Comment:
So we accept OrtValue from the user, serialize it to TensorProto and then deserialize it back to OrtValue at a later stage. Would this not cause the memory to be allocated twice for these initializers - once by the user and second time by the framework? or am I missing something?

Reply:
The user must create their own buffers, this is where they do their processing such as decryption. TensorProto cannot point to the user defined buffers. When the initializers are converted to Tensors the memory in the graph is freed. This has a tradeoff of minimizing code changes.  Another trade off is that the user does not have to copy initializers to the GPU and no special treatment required as it is all handled by the standard mechanism.

Reply:
Hmm, this might be worth calling out in the API header as it's not obvious esp. for users who are sensitive to memory usage. Since this particular API is required only in the case of large models (otherwise there wouldn't be a need to externalize the initializers), this tradeoff is an important one.

Reply:
Is this even a viable solution given it means there's 2x the memory required and we're talking about models so large they need to store initializers in external data?